src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/jobs/LicenceActivationService.kt

```diff
@@ -0,0 +1,94 @@
+    val (eligibleLicences, ineligibleLicences) = determineLicenceEligibility(potentialLicences)
+    val licencesToActivate = determineEligibleLicencesToActivate(eligibleLicences)
+
+    if (licencesToActivate.any()) {
+      licenceService.activateLicences(licencesToActivate)
+    }
+    if (ineligibleLicences.any()) {
+      licenceService.inactivateLicences(ineligibleLicences)
+    }
+  }
+
+  private fun determineLicenceEligibility(licences: List<Licence>): Pair<List<Licence>, List<Licence>> {
+    val bookingIds = licences.map { it.bookingId!! }
+    val hdcStatuses = prisonApiClient.hdcStatuses(bookingIds)
+    val approvedHdcBookingIds = hdcStatuses.filter { it.approvalStatus == "APPROVED" }.map { it.bookingId }
+    val (ineligibleLicences, eligibleLicences) = licences.partition {
+      approvedHdcBookingIds.contains(it.bookingId)
+    }
+    return Pair(eligibleLicences, ineligibleLicences)
+  }
+
+  private fun determineEligibleLicencesToActivate(licences: List<Licence>): List<Licence> {
+    val (iS91Licences, standardLicences) = filterLicencesIntoTypes(licences)
+    val standardLicencesPrisoners =
+      prisonerSearchApiClient.searchPrisonersByBookingIds(standardLicences.map { it.bookingId!! })
+
+    val iS91LicencesToActivate = iS91Licences.filter { it.passedIS91ReleaseDate() }
+    val standardLicencesToActivate =
+      standardLicences.filter { it.standardLicenceForActivation(standardLicencesPrisoners) }
+
+    return iS91LicencesToActivate + standardLicencesToActivate
+  }
+
+  private fun filterLicencesIntoTypes(licences: List<Licence>): Pair<List<Licence>, List<Licence>> {
+    val licenceBookingIds = licences.map { it.bookingId!! }
+    val iS91AndExtraditionBookingIds = prisonApiClient.getIS91AndExtraditionBookingIds(licenceBookingIds)
+    val (iS91AndExtraditionLicences, standardLicences) =
+      licences.partition { iS91AndExtraditionBookingIds.contains(it.bookingId) }
+    return Pair(iS91AndExtraditionLicences, standardLicences)
+  }
+
+  private fun Licence.passedIS91ReleaseDate(): Boolean {
+    if (this.conditionalReleaseDate == null) {
+      return false
+    }
+    // If ARD within CRD minus 4 days and CRD (inclusive), use ARD
+    val releaseDate = if (this.actualReleaseDate != null &&
+      !this.actualReleaseDate.isBefore(this.conditionalReleaseDate.minusDays(4)) &&
+      !this.actualReleaseDate.isAfter(this.conditionalReleaseDate)
+    ) {
+      this.actualReleaseDate
+    } else {
+      this.conditionalReleaseDate
+    }
+
+    return releaseDate <= LocalDate.now()
+  }
+
+  private fun Licence.standardLicenceForActivation(prisoners: List<PrisonerSearchPrisoner>): Boolean {
+    val prisoner: PrisonerSearchPrisoner = prisoners.first {
+      it.prisonerNumber == this.nomsId
+    }
+    if (this.actualReleaseDate != null &&
+      this.actualReleaseDate <= LocalDate.now() &&
+      prisoner.status?.startsWith("INACTIVE") == true
+    ) {
```

Comment:
` if (a) { return true } else { return false } ` => `return a`

Reply:
If I understand correctly, you mean I should just drop the `== true`?
The potential null-state of `prisoner.status` meant that doing `prisoner.status?.startsWith("INACTIVE")` without an explicit `true` check threw an error. Specifically a type mismatch as the condition was expecting a boolean but could receive a `null`

Reply:
I meant the whole if/else  - 

you can just:
```
return this.actualReleaseDate != null 
  && this.actualReleaseDate <= LocalDate.now() 
  && prisoner.status?.startsWith("INACTIVE") == true

```

Reply:
Oh yeah of course, I didn't even think about that!